src/checks/substitution_key.rs

```diff
@@ -0,0 +1,184 @@
+use crate::checks::Check;
+use crate::common::*;
+
+pub(crate) struct SubstitutionKeyChecker<'a> {
+    name: &'a str,
+    template: &'a str,
+}
+
+impl Default for SubstitutionKeyChecker<'_> {
+    fn default() -> Self {
+        Self {
+            name: "SubstitutionKey",
+            template: "The {} key is not assigned properly",
+        }
+    }
+}
+
+impl Check for SubstitutionKeyChecker<'_> {
+    fn run(&mut self, line: &LineEntry) -> Option<Warning> {
+        let mut value = match line.get_value().map(str::trim) {
+            Some(value) if !value.starts_with('\'') => value,
+            _ => return None,
+        };
+
+        let is_escaped =
+            |prefix: &str| prefix.chars().rev().take_while(|ch| *ch == '\\').count() % 2 == 1;
+
+        // Checks if keys used in value have both '{' '}' or neither
+        while let Some(index) = value.find('$') {
+            let prefix = &value[..index];
+            let raw_key = &value[index + 1..];
+
+            // Separate initial key from the rest
+            let (initial_key, rest) = raw_key
+                .find(|c: char| c == '$')
+                .map(|i| (&raw_key[..i], &raw_key[i..]))
+                .unwrap_or_else(|| (raw_key, ""));
+
+            // Ensure if key starts with a '{' that it ends with a '}', else set
+            // bad_substitution to true. Also set key to the proper substitution key.
+            let (key, bad_substitution) = match initial_key.find(|c: char| c == '{') {
+                Some(0) => initial_key
+                    .find(|c: char| c == '}')
+                    .map(|i| (&initial_key[1..i], false))
+                    .unwrap_or_else(|| (&initial_key, true)),
+                _ => initial_key
+                    .find(|c: char| !c.is_ascii_alphanumeric() && c != '_')
+                    .map(|i| (&initial_key[..i], false))
+                    .unwrap_or_else(|| (initial_key, false)),
+            };
+
+            if (bad_substitution
+                || key
+                    .chars()
+                    .any(|c: char| !c.is_ascii_alphanumeric() && c != '_'))
+                && !is_escaped(prefix)
```

Comment:
I think this can be minimized since the value has already been pre-truncated 🤔 

```suggestion
            let has_start_brace = initial_key.starts_with('{');
            let has_end_brace = initial_key.contains('}');
            let is_incorrect_substitution = has_start_brace ^ has_end_brace;

            if is_incorrect_substitution && !is_escaped(prefix) {
```

Reply:
Ok, seems like that would be nearly the same as before last commit though, but will fail in some of the cases.

Reply:
> Ok, seems like that would be nearly the same as before last commit though, but will fail in some of the cases.

In general, yes, but in previous versions we did not truncate the value to the next `$`.

Reply:
Take a look at the [slightly updated version](https://github.com/dotenv-linter/dotenv-linter/pull/414/files#r615419849)